### PR TITLE
On X11, use events modifiers to detect state

### PR DESCRIPTION
**Pull request by kchibisov**
While there's a separate event to deliver modifiers for keyboard, unfortunately, it's not even remotely reflects the modifiers state.

Thus use events along side regular modifier updates to correctly detect the state. Also, apply the modifiers from the regular key event by converting their state to xkb modifiers state.

Links: https://github.com/alacritty/alacritty/issues/7549
Closes: #3388

- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users

### CODE REVIEW

**Comment by kchibisov** [comment]
@daxpedda @madsmtm @notgull do you have a suggestion what to do with the CI?

**Comment by daxpedda** [comment]
This happens because of missing MSRV support in Cargo dependency selection: https://github.com/rust-lang/cargo/issues/9930.

One way to go here is just to fix the dependency version in the CI with `cargo update -p ahash --precise version`.
Alternatively we can also follow [Cargo's recommendation](https://doc.rust-lang.org/cargo/guide/continuous-integration.html#verifying-rust-version) and use [`cargo-hack`](https://crates.io/crates/cargo-hack).

**Comment by kchibisov** [comment]
Can't we have a `Cargo.lock` with precise versions? And then just exclude it from publishing.

**Comment by daxpedda** [comment]
> Can't we have a `Cargo.lock` with precise versions? And then just exclude it from publishing.

I'm very strongly against committing a `Cargo.lock` file **unless** the issues I mentioned in https://github.com/rust-windowing/winit/issues/3085 would be addressed as well.

**Comment by kchibisov** [comment]
>I'm very strongly against committing a Cargo.lock file unless the issues I mentioned in https://github.com/rust-windowing/winit/issues/3085 would be addressed as well.

It's a default behavior for libraries now even. But I don't see what issue it creates and why you need dependabot for that. It only solves issues, like having _working bisect_. 


It's also worth saying that I'd rather test min version than latest deps, since if semver breaks it's not our issue, and if winit can't work correctly without version `0.8.N+1` we should force it in `Cargo.toml`, otherwise you have broken build which you don't notice because you pull latest and things magically work.

**Comment by kchibisov** [comment]
I've used the `--precise` for now.

File: src/platform_impl/linux/x11/event_processor.rs [hunk 1]
```diff
@@ -1527,42 +1548,148 @@ impl EventProcessor {
...
+        }
+        if let Some(mod3) = mods_indices.mod3.filter(|_| ModMask::M3.intersects(state)) {
+            depressed |= 1 << mod3;
+        }
+        if let Some(logo) = mods_indices.logo.filter(|_| ModMask::M4.intersects(state)) {
+            depressed |= 1 << logo;
+        }
+        if let Some(mod5) = mods_indices.mod5.filter(|_| ModMask::M5.intersects(state)) {
+            depressed |= 1 << mod5;
+        }
```
Comment (notgull) [inline src/platform_impl/linux/x11/event_processor.rs:1688]: This code is repetitive and could be deduplicated with a macro

**Review by notgull — APPROVED** [review]
Minor quibbles but I'd rather get this merged and fix the quibbles later.

**Pull request merged**